### DISCUSSION TITLE
Prevent `mason-lspconfig` from starting up a conflicting rust_analyzer LSP client

### DISCUSSION THREAD

**Discussion by OMGeeky** [mrcjkb/rustaceanvim#94] [Q&A] [answered]
With a basically default LazyVim config I get the following error with the `:checkhealt rustaceanvim` while having a rust file open.

> `- ERROR lspconfig.rust_analyzer has been setup. This will likely lead to conflicts with the rustaceanvim LSP client.`

I have not setup anything to my knowledge that should setup rust_analyzer besides this and I can't figure out whats wrong...

Maybe you can take a look at my config here: [OMGeeky/nvim-config-1](https://github.com/OMGeeky/nvim-config-1)

I am quite inexperienced with nvim so it might just be something stupid but I just can't figure it out. Thanks for any help you can provide.

**Comment by mrcjkb** [comment, marked as answer]
Hey 👋 

I'm not familiar with the internals of LazyVim, and I can't see anything wrong in your config.

The health check searches for rust filetype autocommands created by lspconfig and reports the error if it finds one.

You could try opening a rust file and running `:LspInfo`. It should tell you if more than one client is attached.
If that's the case, then you can try explicitly disabling the `rust_analyzer` configuration in lspconfig with the following LazyVim snippet:

```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      setup = {
        rust_analyzer = function()
          return true 
        end,
      },
    },
  }
```

Or, more generally (for anyone not using LazyVim):

```lua
require("mason-lspconfig").setup {
    automatic_enable = {
        exclude = {
            "rust_analyzer",
        }
    }
}
```

**Reply by OMGeeky** [reply]
Using `:LspInfo` showed that there were indeed two things running. one with the name `rust-analyzer` and one with the name `rust_analyzer` and the snippet you provided solved that problem so now there is only one left (the `rust-analyzer` one).

**Reply by kaiserbh** [reply]
Hi @mrcjkb, so the fix "works" for me as well it does stop spawning mason `rust_analyzer` but I realised that my LSP is completely stopped as well, so I can't jump into function reference or use any other lsp features. Is it possible to use mason `rust_analyzer` instead of spawning it's own `rust-analyzer`?

It might be the way I am doing it or my config though so I will just paste my conf here for setting up the lsp I commented the code that disables mason's `rust_analyzer` .


```prolog
mason_lspconfig.setup_handlers({
  function(server_name)
    require('lspconfig')[server_name].setup({
      capabilities = capabilities,
      on_attach = on_attach,
      settings = servers[server_name],
      filetypes = (servers[server_name] or {}).filetypes,
    })
  end,

  --  Disable rust_analyzer that way it doesn't create two instance when using rustceanvim plugin https://github.com/mrcjkb/rustaceanvim/discussions/94#discussioncomment-7813716
  -- ['rust_analyzer'] = function() end,
})
```

**Reply by mrcjkb** [reply]
@KaiserBh

> Is it possible to use mason rust_analyzer instead of spawning it's own rust-analyzer?

No.

What's the output of `:checkhealth rustaceanvim`?
This plugin won't start the LSP client if it can't find rust-analyzer on the `PATH`.
Also, are you setting up the keymaps for rustaceanvim?

**Reply by kaiserbh** [reply]
Heya @mrcjkb , thanks for the quick reply.

Hmm I think the LSP is working as intended, it's probably the keymaps I thought it will overtake or use my default on_attach LSP keymaps, do I need to use a separate keymaps for them or can I maybe use my on_attach function that I made for example this:

```lua
local on_attach = function(_, bufnr)
  local nmap = function(keys, func, desc)
    if desc then
      desc = 'LSP: ' .. desc
    end

    vim.keymap.set('n', keys, func, { buffer = bufnr, desc = desc })
  end

  nmap('<leader>rn', vim.lsp.buf.rename, '[R]e[n]ame')
  nmap('<leader>ca', vim.lsp.buf.code_action, '[C]ode [A]ction')

  nmap('gd', require('telescope.builtin').lsp_definitions, '[G]oto [D]efinition')
  nmap('gr', require('telescope.builtin').lsp_references, '[G]oto [R]eferences')
  nmap('gI', require('telescope.builtin').lsp_implementations, '[G]oto [I]mplementation')
  nmap('<leader>D', require('telescope.builtin').lsp_type_definitions, 'Type [D]efinition')
  nmap('<leader>ds', require('telescope.builtin').lsp_document_symbols, '[D]ocument [S]ymbols')
  nmap('<leader>ws', require('telescope.builtin').lsp_dynamic_workspace_symbols, '[W]orkspace [S]ymbols')

  -- See `:help K` for why this keymap
  nmap('K', vim.lsp.buf.hover, 'Hover Documentation')
  nmap('<C-k>', vim.lsp.buf.signature_help, 'Signature Documentation')

  -- Lesser used LSP functionality
  nmap('gD', vim.lsp.buf.declaration, '[G]oto [D]eclaration')
  nmap('<leader>wa', vim.lsp.buf.add_workspace_folder, '[W]orkspace [A]dd Folder')
  nmap('<leader>wr', vim.lsp.buf.remove_workspace_folder, '[W]orkspace [R]emove Folder')
  nmap('<leader>wl', function()
    print(vim.inspect(vim.lsp.buf.list_workspace_folders()))
  end, '[W]orkspace [L]ist Folders')

  -- Create a command `:Format` local to the LSP buffer
  vim.api.nvim_buf_create_user_command(bufnr, 'Format', function(_)
    vim.lsp.buf.format()
  end, { desc = 'Format current buffer with LSP' })
end

-- document existing key chains
require('which-key').register({
  ['<leader>c'] = { name = '[C]ode', _ = 'which_key_ignore' },
  ['<leader>d'] = { name = '[D]ocument', _ = 'which_key_ignore' },
  ['<leader>h'] = { name = 'More git', _ = 'which_key_ignore' },
  ['<leader>r'] = { name = '[R]ename', _ = 'which_key_ignore' },
  ['<leader>s'] = { name = '[S]earch', _ = 'which_key_ignore' },
  ['<leader>w'] = { name = '[W]orkspace', _ = 'which_key_ignore' },
})
```

**Reply by mrcjkb** [reply]
Yes, you can reuse your `on_attach`. See the tip box in the readme's [Quick Setup](https://github.com/mrcjkb/rustaceanvim?tab=readme-ov-file#quick-setup) section 😀

**Reply by utensil** [reply]
It seems that none of the solutions work for me, after adding all these configs, tried some variants, e.g. the general one:

<img width="957" alt="image" src="https://github.com/user-attachments/assets/be0a20d1-e9ed-44ae-9dfa-067162d3d2e6">

even after I manually run `MasonUninstall rust-analyzer`, I'll still be prompt that rust_analyzer installation in progress when opening a rust file,

<img width="440" alt="image" src="https://github.com/user-attachments/assets/2f79ebe7-8a7c-4881-9281-f02549d9bcb5">

then two instances come back:

<img width="523" alt="image" src="https://github.com/user-attachments/assets/04d72c09-b727-42cb-9203-6dfc5bc181c9">

<img width="1279" alt="image" src="https://github.com/user-attachments/assets/b84f8d33-1b19-433d-9fe9-035940a6d8a3">

<img width="1090" alt="image" src="https://github.com/user-attachments/assets/cc5c6861-2d4d-4561-b72b-977580a275c2">

Running the help gives me some errors:

<img width="1203" alt="image" src="https://github.com/user-attachments/assets/ee7c0a35-4cdf-463d-91d1-b9b0762a15b5">

But after a few enters I managed to read it. The solution inside is the same as this answer.

I have also tried a few times of a clean LunarVim installation.

**Reply by mrcjkb** [reply]
Perhaps you are setting up lspconfig.rust_analyzer somewhere else?
There's not much I can do to help here. You'll have to bisect your plugins and your config.

**Reply by utensil** [reply]
Thanks for your reply. My config is [very short](https://github.com/utensil/forest/blob/cf2a2ce4c4635d9e5797227bb1f0006f0b3c5cad/config.lua) and nothing seems suspicious, it could be somewhere inside LunarVim, but I can't bisect it.

**Reply by mrcjkb** [reply]
I'm afraid I don't have the capacity to support all the Neovim distros out there.
It's best you ask in the LunarVim discussions.

**Reply by utensil** [reply]
Thanks, I just figured out why:

<img width="1016" alt="image" src="https://github.com/user-attachments/assets/9feb27f8-0da3-46ee-90d3-5f1296169180">


So my removing and reinstalling of LunarVim is not thorough, and `:LvimCacheReset` must be run for my config line

```
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })
```

to take effect.

Now only one rust analyzer is running (LSP progress updated by fidget.nvim, which showed two rust_analyzer running `cargo check` simultaneously before the fix):

<img width="648" alt="image" src="https://github.com/user-attachments/assets/07dc6adf-9c96-46ad-a6d1-e463c545b5bc">

**Reply by n-prat** [reply]
For anyone finding that in late 2025, the way to do that is now [this](https://github.com/mason-org/mason-lspconfig.nvim#automatically-enable-installed-servers) cf `To exclude certain servers from being enabled:` eg

```
require("mason-lspconfig").setup {
    automatic_enable = {
        exclude = {
            "rust_analyzer",
        }
    }
}
```

**Reply by mrcjkb** [reply]
@n-prat thanks 🙏
I've updated my answer.
